File: library/ilib/web/ffi/web_ffi.java
import java.awt.Desktop;
import java.net.URI;

class WebLib {
    public static void open(String link) {
        openURL("https://" + link);
    }

    public static void fileOpen(String file) {
        openURL("file://" + file);
    }

    public static void popen(String rawLink) {
        openURL(rawLink);
    }

    public static void ropen(String identifier) {
        openURL("https://" + identifier + ".com");
    }

    public static void browser() {
        openURL("https://www.google.com");
    }

    public static boolean pdf(String pdf) {
        if (pdf.endsWith(".pdf")) {
            fileOpen(pdf);
            return true;
        }
        return false;
    }

    public static boolean text(String text) {
        String[] validExts = {".txt", ".md", ".bashrc", ".zshrc"};
        for (String ext : validExts) {
            if (text.endsWith(ext)) {
                fileOpen(text);
                return true;
            }
        }
        return false;
    }

    public static boolean inspect(String program) {
        String[] validExts = {".py", ".java", ".js", ".c", ".cpp", ".v", ".ac", ".s", ".go", ".rs", ".sh", ".html", ".sql"};
        for (String ext : validExts) {
            if (program.endsWith(ext)) {
                fileOpen(program);
                return true;
            }
        }
        return false;
    }

    public static void acPage() {
        openURL("https://aclang.vercel.app");
    }

    public static String pageGet(String url) {
        // HTTP(S) GET returning the body as a string (HttpURLConnection, stdlib).
        // Adds https:// if no scheme; "" on error.
        String full = url.contains("://") ? url : "https://" + url;
        try {
            java.net.HttpURLConnection conn =
                (java.net.HttpURLConnection) new java.net.URL(full).openConnection();
            conn.setRequestMethod("GET");
            conn.setConnectTimeout(15000);
            conn.setReadTimeout(15000);
            conn.setInstanceFollowRedirects(true);
            java.io.InputStream in = conn.getInputStream();
            java.io.ByteArrayOutputStream out = new java.io.ByteArrayOutputStream();
            byte[] buf = new byte[4096];
            int n;
            while ((n = in.read(buf)) != -1) out.write(buf, 0, n);
            in.close();
            return out.toString("UTF-8");
        } catch (Exception e) {
            return "";
        }
    }

    public static void help() {
        String helpText = "=== Web Library Functions ===\n" +
            "web.open(link)           - Opens https://{link}\n" +
            "web.fileOpen(file)       - Opens file://{file}\n" +
            "web.popen(rawLink)       - Opens {rawLink} as-is\n" +
            "web.ropen(search)        - Opens https://{search}.com\n" +
            "web.browser()            - Opens https://www.google.com\n" +
            "web.pdf(pdf)             - Opens file://{pdf}, must be .pdf\n" +
            "web.text(text)           - Opens file://{text}, must be .txt, .md, .bashrc, or .zshrc\n" +
            "web.inspect(program)     - Opens file://{program}, must be source code\n" +
            "web.acPage()             - Opens the AC language website\n" +
            "web.pageGet(link)        - Gets HTML contents of a webpage\n" +
            "web.help()               - Shows this help message";
        System.out.println(helpText);
    }

    private static void openURL(String url) {
        try {
            if (Desktop.isDesktopSupported()) {
                Desktop.getDesktop().browse(new URI(url));
            } else {
                openURLFallback(url);
            }
        } catch (Exception e) {
            openURLFallback(url);
        }
    }

    private static void openURLFallback(String url) {
        try {
            String osName = System.getProperty("os.name").toLowerCase();
            if (osName.contains("win")) {
                Runtime.getRuntime().exec("cmd /c start " + url);
            } else if (osName.contains("mac")) {
                Runtime.getRuntime().exec("open " + url);
            } else if (osName.contains("linux") || osName.contains("unix")) {
                Runtime.getRuntime().exec("xdg-open " + url);
            }
        } catch (Exception e) {
            System.err.println("Could not open URL: " + e.getMessage());
        }
    }
}

class web {
    public static void open(String link) { WebLib.open(link); }
    public static void file_open(String file) { WebLib.fileOpen(file); }
    public static void popen(String rawLink) { WebLib.popen(rawLink); }
    public static void ropen(String identifier) { WebLib.ropen(identifier); }
    public static void browser() { WebLib.browser(); }
    public static long pdf(String pdf) { return WebLib.pdf(pdf) ? 1L : 0L; }
    public static long text(String text) { return WebLib.text(text) ? 1L : 0L; }
    public static long inspect(String program) { return WebLib.inspect(program) ? 1L : 0L; }
    public static void ac_page() { WebLib.acPage(); }
    public static String page_get(String url) { return WebLib.pageGet(url); }
    public static String help() {
        WebLib.help();
        return "=== Web Library Functions ===";
    }
}

File: library/ilib/web/ffi/web_ffi.rs
use std::process::Command;

pub struct WebLib;

impl WebLib {
    pub fn open(&self, link: &str) {
        Self::open_url(&format!("https://{}", link));
    }

    pub fn file_open(&self, file: &str) {
        Self::open_url(&format!("file://{}", file));
    }

    pub fn popen(&self, raw_link: &str) {
        Self::open_url(raw_link);
    }

    pub fn ropen(&self, identifier: &str) {
        Self::open_url(&format!("https://{}.com", identifier));
    }

    pub fn browser(&self) {
        Self::open_url("https://www.google.com");
    }

    pub fn pdf(&self, pdf: &str) -> i64 {
        if pdf.ends_with(".pdf") {
            self.file_open(pdf);
            1
        } else {
            0
        }
    }

    pub fn text(&self, text: &str) -> i64 {
        let valid_exts = [".txt", ".md", ".bashrc", ".zshrc"];
        if valid_exts.iter().any(|ext| text.ends_with(ext)) {
            self.file_open(text);
            1
        } else {
            0
        }
    }

    pub fn inspect(&self, program: &str) -> i64 {
        let valid_exts = [".py", ".java", ".js", ".c", ".cpp", ".v", ".ac", ".s", ".go", ".rs", ".sh", ".html", ".sql"];
        if valid_exts.iter().any(|ext| program.ends_with(ext)) {
            self.file_open(program);
            1
        } else {
            0
        }
    }

    pub fn ac_page(&self) {
        Self::open_url("https://aclang.vercel.app");
    }

    pub fn page_get(&self, url: &str) -> String {
        // HTTP(S) GET via curl (no external crate). Adds https:// if no scheme; only allows
        // safe http(s) URLs (blocks shell/arg injection). Returns "" on error.
        let u = if url.contains("://") { url.to_string() } else { format!("https://{}", url) };
        let ok = (u.starts_with("http://") || u.starts_with("https://"))
            && !u.chars().any(|c| c.is_control() || " '\"`\\;|&$<>()*?".contains(c));
        if !ok { return String::new(); }
        match Command::new("curl").args(["-sL", "--max-time", "15", "--", &u]).output() {
            Ok(o) => String::from_utf8_lossy(&o.stdout).into_owned(),
            Err(_) => String::new(),
        }
    }

    pub fn help(&self) -> String {
        let help_text = r#"=== Web Library Functions ===
web.open(link)           - Opens https://{link}
web.file_open(file)      - Opens file://{file}
web.popen(raw_link)      - Opens {raw_link} as-is
web.ropen(search)        - Opens https://{search}.com
web.browser()            - Opens https://www.google.com
web.pdf(pdf)             - Opens file://{pdf}, must be .pdf
web.text(text)           - Opens file://{text}, must be .txt, .md, .bashrc, or .zshrc
web.inspect(program)     - Opens file://{program}, must be source code
web.ac_page()            - Opens the AC language website
web.page_get(link)       - Gets HTML contents of a webpage
web.help()               - Shows this help message"#;
        println!("{}", help_text);
        help_text.to_string()
    }

    fn open_url(url: &str) {
        #[cfg(target_os = "windows")]
        {
            let _ = Command::new("cmd")
                .args(&["/C", "start", url])
                .output();
        }
        #[cfg(target_os = "macos")]
        {
            let _ = Command::new("open").arg(url).output();
        }
        #[cfg(target_os = "linux")]
        {
            let _ = Command::new("xdg-open").arg(url).output();
        }
    }
}

pub static web: WebLib = WebLib;
